### PR TITLE
Added the join and Unjoin API and get APIs

### PR DESCRIPTION
**Pull request by ruksana2808**

### CODE REVIEW

File: src/main/java/com/igot/cb/community/controller/CommunityController.java [hunk 1]
```diff
@@ -45,4 +46,24 @@ public ResponseEntity<ApiResponse> update(@RequestBody JsonNode communityDetails
...
     }
+
+    @PutMapping("/join")
+    public ResponseEntity<ApiResponse> join(@RequestBody Map<String, Object> request,
+        @RequestHeader(Constants.X_AUTH_TOKEN) String authToken) {
+        ApiResponse response = communityManagementService.joinAndUnjoinCommunity(request, authToken);
+        return new ResponseEntity<>(response, response.getResponseCode());
+    }
+
+    @GetMapping("/read/user/joined")
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/community/controller/CommunityController.java]: /user/communities -- should be the endpoint

File: src/main/java/com/igot/cb/community/controller/CommunityController.java [hunk 2]
```diff
@@ -45,4 +46,24 @@ public ResponseEntity<ApiResponse> update(@RequestBody JsonNode communityDetails
...
+    public ResponseEntity<ApiResponse> readJoin(@RequestHeader(Constants.X_AUTH_TOKEN) String authToken) {
+        ApiResponse response = communityManagementService.communitiesJoinedByUser(authToken);
+        return new ResponseEntity<>(response, HttpStatus.OK);
+    }
+
+    @GetMapping("/joined/users/{communityId}")
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/community/controller/CommunityController.java]: /community/listuser/{communityId}

File: src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java [hunk 3]
```diff
@@ -284,4 +290,152 @@ public ApiResponse update(JsonNode communityDetails, String authToken) {
...
+    public ApiResponse joinAndUnjoinCommunity(Map<String, Object> request, String authToken) {
+        log.info("CommunityEngagementService:joinAndUnjoinCommunity:joining");
+        ApiResponse response = ProjectUtil.createDefaultResponse(Constants.API_COMMUNITY_UPDATE);
+        String error = validateJoinPayload(request);
+        if (StringUtils.isNotBlank(error)) {
+            return returnErrorMsg(error, HttpStatus.BAD_REQUEST, response);
+        }
+
+        try {
+            String userId = accessTokenValidator.verifyUserToken(authToken);
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java:298]: why this method returns "Unauthorized" ?? just return empty or return proper user id value.

File: src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java [hunk 4]
```diff
@@ -284,4 +290,152 @@ public ApiResponse update(JsonNode communityDetails, String authToken) {
         }
     }
 
+    @Override
+    public ApiResponse joinAndUnjoinCommunity(Map<String, Object> request, String authToken) {
+        log.info("CommunityEngagementService:joinAndUnjoinCommunity:joining");
+        ApiResponse response = ProjectUtil.createDefaultResponse(Constants.API_COMMUNITY_UPDATE);
+        String error = validateJoinPayload(request);
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java]: Make sue accessToken validated first

File: src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java [hunk 5]
```diff
@@ -284,4 +290,152 @@ public ApiResponse update(JsonNode communityDetails, String authToken) {
...
+        try {
+            String userId = accessTokenValidator.verifyUserToken(authToken);
+            if (StringUtils.isBlank(userId) || userId.equalsIgnoreCase(
+                Constants.UNAUTHORIZED_USER)) {
+                response.getParams().setErrMsg(Constants.USER_ID_DOESNT_EXIST);
+                response.setResponseCode(HttpStatus.BAD_REQUEST);
+                return response;
+            }
+            String communityId = (String) request.get(Constants.COMMUNITY_ID);
+            Optional<CommunityEntity> optComment = communityEngagementRepository.findByCommunityIdAndIsActive(
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java]: use proper variable name

File: src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java [hunk 6]
```diff
@@ -284,4 +290,152 @@ public ApiResponse update(JsonNode communityDetails, String authToken) {
...
+                communityId, true);
+            if (optComment == null || !optComment.isPresent() || optComment.get().getData()
+                .isEmpty()) {
+                response.setResponseCode(HttpStatus.BAD_REQUEST);
+                response.getParams().setErr(error);
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java]: what would be the value for variable error ?

File: src/main/java/com/igot/cb/community/controller/CommunityController.java [hunk 7]
```diff
@@ -45,4 +46,24 @@ public ResponseEntity<ApiResponse> update(@RequestBody JsonNode communityDetails
         ApiResponse response = communityManagementService.update(communityDetails, authToken);
         return new ResponseEntity<>(response, response.getResponseCode());
     }
+
+    @PutMapping("/join")
+    public ResponseEntity<ApiResponse> join(@RequestBody Map<String, Object> request,
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/community/controller/CommunityController.java:51]: this implementation is not correct.... 
UI will have the details while loading that the logged in user is joined or not joined... 
so it has to be join operation only and do define one more endpoint to leave from community... 
you can have service method which takes param -- just pass it is add operation or leave operation

File: src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java [hunk 8]
```diff
@@ -284,4 +290,152 @@ public ApiResponse update(JsonNode communityDetails, String authToken) {
...
+                return response;
+            }
+            Map<String, Object> propertyMap = new HashMap<>();
+            propertyMap.put(Constants.USER_ID, userId);
+            propertyMap.put(Constants.CommunityId, communityId);
+            List<Map<String, Object>> userCommunityDetails = cassandraOperation.getRecordsByPropertiesWithoutFiltering(
+                Constants.KEYSPACE_SUNBIRD, Constants.USER_COMMUNITY_TABLE, propertyMap, null, 1);
+            if (CollectionUtils.isEmpty(userCommunityDetails)) {
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java:322]: this implementation is not correct... based on incoming operation you need to either join the community or leave from community... 
but here if the record exist - it is removing the user and if record not exist it adds the user... 

what will happen if user clicks twice or the api called twice by mistake... ?

File: src/main/java/com/igot/cb/transactional/cassandrautils/CassandraOperationImpl.java [hunk 9]
```diff
@@ -201,6 +202,24 @@ public Map<String,Object> updateRecord(
         return response;
     }
 
+    @Override
+    public void deleteRecord(String keyspaceName, String tableName, Map<String, Object> compositeKeyMap) {
```
Comment (karthik-tarento) [inline src/main/java/com/igot/cb/transactional/cassandrautils/CassandraOperationImpl.java]: there should not be any delete operation in cassandra... 
we should do only soft delete -- that is updating a column in that table mentioning record is inactive.

**Review by karthik-tarento — CHANGES_REQUESTED** [review]

File: src/main/java/com/igot/cb/community/controller/CommunityController.java (hunk 1 above)
Comment (ruksana2808) [reply src/main/java/com/igot/cb/community/controller/CommunityController.java]: done

File: src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java (hunk 3 above)
Comment (ruksana2808) [reply src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java:298]: done

File: src/main/java/com/igot/cb/community/controller/CommunityController.java (hunk 7 above)
Comment (ruksana2808) [reply src/main/java/com/igot/cb/community/controller/CommunityController.java:51]: Made change to handle join and un-join in separate API

File: src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java (hunk 8 above)
Comment (ruksana2808) [reply src/main/java/com/igot/cb/community/service/impl/CommunityManagementServiceImpl.java:322]: Implemented soft delete

File: src/main/java/com/igot/cb/transactional/cassandrautils/CassandraOperationImpl.java (hunk 9 above)
Comment (ruksana2808) [reply src/main/java/com/igot/cb/transactional/cassandrautils/CassandraOperationImpl.java]: Implemented soft delete

**Pull request merged**